vinci1it2000/formulas
# Jbdev

## Status
**READY/IN DEVELOPMENT/HOLD**
Versioned as V1.1.0 from master HEAD. Merged into dev HEAD.
Manual merging of test/test_files/test.xlsx amounted to updating IFS row 8 in LOGIC sheet.
Links in B17 & B18 of CORE sheet in text.xlsx had to be updated to local absolute paths for
tests to succeed. 

## Description
Added _XLFN.IFS() function support with behaviour matching Microsoft 365 Excel 2016.
Parser and Excel tests for IFS() are provided.

Updated OR() error responses to match behaviour in Microsoft 365 Excel 2016 in test.xlsx LOGIC sheet.
File test.xlsx is saved in Excel 2016 format.

Added SINGLE() function placeholder to deal with a specific '@' implicit intersection test case for IFS().
Excel 2016 changes some '@' operators to SINGLE() behind the scenes, as read by openpyxl,
but still displays '@' on-screen for these cases.
SINGLE() is now implemented but only to return #VALUE! errors.
The only package tests for SINGLE are a parser test and an implicit #VALUE! error test with IFS() in test.xlsx LOGIC sheet cell AF8 saved from Excel 2016.
Neither is '@' implemented, even though it is interchangeable in Excel 2016 with SINGLE().

## Related Issues

Issue | Description
----- | ------------

## Todos
Consider full support for '@' and SINGLE, which requires processor function knowledge of current cell address fed through wrappers.

## Steps to Test or Reproduce
Run IFS() tests added to test.xlsx (in Excel 2016 format) and to parser tests.
'@' tokens appear to be dropped before parsing (openpyxl issue perhaps?)

```sh
git pull --prune
git checkout <feature_branch>
python
...
```
